sql: Add crdb_internal.locality_value builtin function

andy-kimball opened the pull request:
There is currently no easy way to programmatically inspect the locality of the
current node. It appears in crdb_internal.gossip_nodes, but it's hard to work
with because it's JSON, not constant-foldable, and keyed by node id.

This commit adds a new crdb_internal.locality_value builtin function that returns
the value of the locality key given as its argument.

When building geo-distributed applications, this enables a very nice way to
automatically assign the partition key, as illustrated below:

  CREATE TABLE charges (
    region STRING NOT NULL DEFAULT crdb_internal.locality_value('region'),
    id UUID NOT NULL DEFAULT gen_random_uuid(),
    ...

The DEFAULT expression for the region column automatically inserts the region
value from the current node's locality.

It also enables queries like this, that only touch rows in the current region
(which is necessary to avoid cross-region hops):

  SELECT * FROM charges WHERE region = crdb_internal.locality_value('region') AND id = $1

The optimizer folds the function to a constant value, which can then be used
to select an optimal index.

Resolves #37310

Release note (sql change): Adds a new locality_value builtin function that
returns the value of the locality key given as its argument.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37369)
<!-- Reviewable:end -->

andy-kimball commented:
@awoods187, I added you to this since I'm proposing to add a new builtin function, which has product/customer impact. See #37310 for more details on the proposal. I'm looking not just for Eng feedback on the implementation, but also Product feedback on how the builtin works.

knz reviewed (commented):
See my comments https://github.com/cockroachdb/cockroach/issues/37310#issuecomment-490460083

The general thrust is all right, but the idea to use labeled tuples is all wrong. Arrays and `unnest` are the way to go, both from a future extensibility perspective (tuple types must be valid regardless of the node where you run) and from a usability perspective (`unnest` gives you access to relational operators to operate on the locality data, tuples wouldn't give you that).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-:-LeMLDy8-mBQYsQKuk0A:bz308r8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @awoods187, @jordanlewis, @knz, and @RaduBerinde)

___
*[pkg/sql/sem/builtins/builtins.go, line 2763 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-LeMKzqh-nmb6vnl1vcs:-LeMKzqjD27KAT29LDMF:bk29zti) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe7ded3d45f907a3ea0215d622fd1c352f499f7c/pkg/sql/sem/builtins/builtins.go#L2763)):*
> ```Go
> 		tree.Overload{
> 			Types: tree.ArgTypes{},
> 			ReturnType: func(ctx *tree.SemaContext, _ []tree.TypedExpr) *types.T {
> ```

I'd say drop this and have it return fixed type `string[][]` (`array[array[string]]`)
___
*[pkg/sql/sem/builtins/builtins.go, line 2773 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-LeML9cM9wIJlx33hjIB:-LeML9cM9wIJlx33hjIC:b-wxhb5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe7ded3d45f907a3ea0215d622fd1c352f499f7c/pkg/sql/sem/builtins/builtins.go#L2773)):*
> ```Go
> 				datums := make(tree.Datums, len(ctx.Locality.Tiers))
> 				for i := range ctx.Locality.Tiers {
> 					datums[i] = tree.NewDString(ctx.Locality.Tiers[i].Value)
> ```

Make a 2-element array of the key, value pair here
___
*[pkg/sql/sem/builtins/builtins.go, line 2776 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-LeMLB_32KBHaF3uobXK:-LeMLB_32KBHaF3uobXL:b-m0itxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe7ded3d45f907a3ea0215d622fd1c352f499f7c/pkg/sql/sem/builtins/builtins.go#L2776)):*
> ```Go
> 				}
> 				typ := localityReturnType(&ctx.Locality)
> 				return tree.NewDTuple(typ, datums...), nil
> ```

and keep the array here


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
I think we should have these discussions on the original issue, here: https://github.com/cockroachdb/cockroach/issues/37310, since that will be easier to find later on if questions come up.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-:-LeMn1tJCK1frqHUP71k:b-j29j1t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @awoods187, @jordanlewis, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

andy-kimball committed:
sql: Add crdb_internal.locality_value builtin function

There is currently no easy way to programmatically inspect the locality of the
current node. It appears in crdb_internal.gossip_nodes, but it's hard to work
with because it's JSON, not constant-foldable, and keyed by node id.

This commit adds a new crdb_internal.locality_value builtin function that returns
the value of the locality key given as its argument.

When building geo-distributed applications, this enables a very nice way to
automatically assign the partition key, as illustrated below:

  CREATE TABLE charges (
    region STRING NOT NULL DEFAULT crdb_internal.locality_value('region'),
    id UUID NOT NULL DEFAULT gen_random_uuid(),
    ...

The DEFAULT expression for the region column automatically inserts the region
value from the current node's locality.

It also enables queries like this, that only touch rows in the current region
(which is necessary to avoid cross-region hops):

  SELECT * FROM charges WHERE region = crdb_internal.locality_value('region') AND id = $1

The optimizer folds the function to a constant value, which can then be used
to select an optimal index.

Resolves #37310

Release note (sql change): Adds a new locality_value builtin function that
returns the value of the locality key given as its argument.

andy-kimball commented:
I changed this PR to add a `crdb_internal.locality_value` function, as per discussion in #37310.

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-:-LeXF3GF3DznakIP6squ:b-g2a5fb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @awoods187, @jordanlewis, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 18 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37369)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @awoods187, and @jordanlewis)

___
*[pkg/sql/opt/norm/norm_test.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-Lekk3L1CQ1DxKhlptx_:-Lekk3L1CQ1DxKhlptxa:b-iw2cac) ([raw file](https://github.com/cockroachdb/cockroach/blob/29a636bd9bd05aa34e8199bb7089b6c17561b731/pkg/sql/opt/norm/norm_test.go#L117)):*
> ```Go
> 		if props.Category == categorySystemInfo || props.Category == categoryDateAndTime {
> 			switch name {
> 			case "crdb_internal.locality_value":
> ```

Why is this ok to fold? Isn't it possible that we get the same exact query from a different locality?


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37369)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @awoods187, @jordanlewis, @knz, and @rytaft)

___
*[pkg/sql/opt/norm/norm_test.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-Lekk3L1CQ1DxKhlptx_:-LelqWyx5Fvhn4Eli3K9:bybz8au) ([raw file](https://github.com/cockroachdb/cockroach/blob/29a636bd9bd05aa34e8199bb7089b6c17561b731/pkg/sql/opt/norm/norm_test.go#L117)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is this ok to fold? Isn't it possible that we get the same exact query from a different locality?
</blockquote></details>

Should always return the same value for a given key, on a given machine. It doesn't matter how many times it's run, or when it's run. It can only return a different value if the node is restarted with a different locality.
___
*[pkg/sql/sem/builtins/builtins.go, line 2763 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-LeMKzqh-nmb6vnl1vcs:-LelqNiN3mP3fyqV2IVH:b-j2eqnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe7ded3d45f907a3ea0215d622fd1c352f499f7c/pkg/sql/sem/builtins/builtins.go#L2763)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'd say drop this and have it return fixed type `string[][]` (`array[array[string]]`)
</blockquote></details>

Switched up the design, so resolving this.
___
*[pkg/sql/sem/builtins/builtins.go, line 2773 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-LeML9cM9wIJlx33hjIB:-LelqQwO-W86kZVmmi0k:b-j2eqnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe7ded3d45f907a3ea0215d622fd1c352f499f7c/pkg/sql/sem/builtins/builtins.go#L2773)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make a 2-element array of the key, value pair here
</blockquote></details>

Switched up the design, so resolving this.
___
*[pkg/sql/sem/builtins/builtins.go, line 2776 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-LeMLB_32KBHaF3uobXK:-LelqSRJ7g9IIBWiB--l:b-j2eqnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe7ded3d45f907a3ea0215d622fd1c352f499f7c/pkg/sql/sem/builtins/builtins.go#L2776)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

and keep the array here
</blockquote></details>

Switched up the design, so resolving this.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/norm/norm_test.go, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-Lekk3L1CQ1DxKhlptx_:-LesjjMp6iHfwgs6DpZ3:b-wsvasg) ([raw file](https://github.com/cockroachdb/cockroach/blob/29a636bd9bd05aa34e8199bb7089b6c17561b731/pkg/sql/opt/norm/norm_test.go#L117)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should always return the same value for a given key, on a given machine. It doesn't matter how many times it's run, or when it's run. It can only return a different value if the node is restarted with a different locality.
</blockquote></details>

Thanks for the explanation!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37369#-:-Lesjr-v7vnK2X1ISwg1:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @awoods187, @jordanlewis, and @knz)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+
